benchmarks/src/bench/json_multi_export.rs

```diff
@@ -4,106 +4,117 @@ use std::{
+                }
+
+                let ok: anyhow::Result<()> = Ok(());
+                ok
+            });
+        }
 
-            for mut doc in docs {
-                doc.insert("file", i as i32);
-                coll.insert_one(doc, None)?;
-            }
+        while let Some(result) = tasks.next().await {
+            println!("done!");
+            result?;
         }
 
-        Ok(JsonMultiExportBenchmark {
-            db,
-            coll,
-            num_threads: options.num_threads,
-        })
+        Ok(JsonMultiExportBenchmark { db, coll })
     }
 
-    fn do_task(&self) -> Result<()> {
-        let mut num_each = TOTAL_FILES / self.num_threads;
-        let remainder = TOTAL_FILES % self.num_threads;
-        if remainder != 0 {
-            num_each += 1;
-        }
+    async fn do_task(&self) -> Result<()> {
+        let mut tasks = FuturesUnordered::new();
 
-        let mut downloaded_files = 0;
-        let mut threads = Vec::new();
-        while downloaded_files < TOTAL_FILES {
-            let num_files = std::cmp::min(TOTAL_FILES - downloaded_files, num_each);
+        for i in 0..TOTAL_FILES {
             let coll_ref = self.coll.clone();
             let path = std::env::temp_dir();
 
-            let thread = std::thread::spawn(move || {
+            tasks.push(async move {
                 // Note that errors are unwrapped within threads instead of propagated with `?`.
-                // While we could set up a channel to send errors back to main thread, this is a lot
-                // of work for little gain since we `unwrap()` in main.rs anyway.
-                for i in downloaded_files..downloaded_files + num_files {
-                    let file_name = path.join(format!("ldjson{:03}.txt", i));
-                    let mut file = OpenOptions::new()
-                        .create(true)
-                        .write(true)
-                        .open(&file_name)
-                        .unwrap();
-
-                    let cursor = coll_ref
-                        .find(Some(doc! { "file": i as i32 }), None)
-                        .unwrap();
-
-                    for doc in cursor {
-                        writeln!(file, "{}", doc.unwrap().to_string()).unwrap();
+                // While we could set up a channel to send errors back to main thread, this is a
+                // lot of work for little gain since we `unwrap()` in
+                // main.rs anyway.
+                let file_name = path.join(format!("ldjson{:03}.txt", i));
+                let mut file = OpenOptions::new()
+                    .create(true)
+                    .write(true)
+                    .open(&file_name)
+                    .unwrap();
+
+                let mut cursor = coll_ref
+                    .find(Some(doc! { "file": i as i32 }), None)
+                    .await
+                    .unwrap();
+
+                let (sender, mut receiver) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
This is the one place in the benchmarks where the conversion to async required a bit of refactoring. The naive implementation of just using `spawn_blocking` like elsewhere doesn't pass the borrow checker due to the fact that the `File` needs to be moved into the `spawn_blocking` closure in order to be passed to `writeln!`. However, this isn't possible due to needing to invoke `writeln!` multiple times in a loop. This could be mitigated by using the runtime-specific async filesystem APIs, but this would require a decent amount of work due to the fact that tokio 0.2.x (the version the driver still supports) does not use the same async read/write traits that async-std uses, so we'd have to abstract over that. The workaround here is to do all of the async driver work in a separate task, then send the results to a blocking task through a channel to be written out. While this might be less performant than using the runtime-specific filesystem APIs, that optimization would only be in the benchmarks themselves and wouldn't actually change anything for end users, so I didn't think it was worth spending time on right now. If later we want to try to squeeze every ounce of performance out for these benchmarks, we can come back and switch to the runtime APIs.

Reply:
Cross driver comparisons of benchmarks are part of the value of running a unified set though, so I think its pretty important we try to cut down on any losses here. The other stuff is fine to leave because it's all in the setup phase. What would be involved in making the requisite abstractions for this test?

Reply:
Basically the same type of stuff we do in the driver for abstracting over tokio/async-std API differences, e.g. with the stream API. I feel somewhat strongly that working to improve the benchmarks themselves but not the driver isn't a priority of this project; right now, we just want to make as big an impact on the actual driver's performance, and putting these benchmarks in evergreen is just a way to measure that. Put another way, time we spend optimizing this benchmark is time we're not spending on stuff like the raw document type, which would be a win for actual users and not just the benchmarks.

Reply:
I agree that optimizing the driver is the highest priority here, but my concern is that if the benchmark suite is artificially slow in a way that affects the results, it makes comparisons against a baseline harder since those baselines come from other drivers which do not have any such overhead. These baseline comparisons are really important, since they guide our expectations and measures of success for our implementations. For instance, we'd like our BSON benchmarks to be roughly on par or better than Java, say, but if they're much slower, we won't be able to tell if its coming from the BSON implementation or the benchmark without further investigation, which also costs time. And when we do fix the runner in the future, we would no longer use our old benchmark data as a baseline either. Another issue is correctness--if our benchmarks aren't using async in the way we expect users to, then it's harder to say if the benchmarks are reflective of real-world performance.

The reason I was wondering how much work would be involved is because if it's a few hours or a day, then it's probably worth it to ensure we have accurate benchmark results, since we likely won't have an opportunity to update these again for a long time. I haven't looked too much into the filesystem APIs, but I can't imagine it's that much different than our socket abstractions which are quite small. Also, we wouldn't have to abstract over the entirety of the filesystem API, but just the pieces that we're actively measuring.

Reply:
Based on offline discussion, all places that are measured by the benchmarking have been updated to use the runtime-specific APIs for filesystem access.